index.bs

```diff
@@ -2086,6 +2130,211 @@ worker=] algorithm:
+
+BaseLogEntry = {
+  level: LogLevel,
+  text: text / null,
+  timestamp: int,
+  ?stackTrace: [*StackFrame],
+}
+
+GenericLogEntry = {
+  BaseLogEntry,
+  type: text,
+}
+
+ConsoleLogEntry = {
+  BaseLogEntry,
+  type: "console",
+  method: text,
+  realm: Realm,
+  args: [*RemoteValue],
+}
+
+JavascriptLogEntry = {
+  BaseLogEntry,
+  type: "javascript",
+}
+
+</pre>
+
+Each log event is represented by a <code>LogEntry</code> object. This has a
+<code>type</code> property which represents the type of log entry added, a
+<code>level</code> property representing severity, a <code>text</code> property
+with the log message string itself, and a <code>timestamp</code> property
+corresponding to the time the log entry was generated. The <code>extra</code>
+property contains additional data specific to the given type.
+
+#### log.StackFrame #### {#types-log-stackframe}
+
+<pre class="cddl local-cddl">
+
+StackFrame = {
+  url: text,
+  functionName: text,
+  lineNumber: int,
+  columnNumber: int,
+}
+
+</pre>
+
+A frame in a stacktrace is represented by a <code>StackFrame</code> object. This
+has a <code>url</code> property, which represents the URL of the script, a
+<code>functionName</code> property which represents the name of the executing
+function, an <code>lineNumber</code> and <code>columnNumber</code> properties,
+which represent the line and column number of the executed code.
+
+The <dfn>current stack trace</dfn> is a representation of the stack of the
+[=running execution context=]. The details of this are unspecified, and so the
+behaviour here is implementation defined, but the general process is as follows:
+
+ 1. Let |stack trace| be a new list.
+
+ 1. For each stack frame |frame| in the stack of the running execution context,
+    starting from the most recently executed frame, run the following steps:
+
+    1. Let |url| be the result of running the [=URL serializer=], given
+       the <a spec=url for=/>URL</a> of |frame|'s associated script resource.
+
+    1. Let |functionName| be the name of |frame|'s associated function.
+
+    1. Let |lineNumber| and |columnNumber| be the on-based line and zero-based
+       column numbers, respectively, of the location in |frame|'s associated
+       script resource corresponding to |frame|.
+
+    1. Let |frame info| be a new map matching the <code>StackFrame</code>
+       production, with the <code>url</code> field set to |url|, the
+       <code>functionName</code> field set to |functionName|, the
+       <code>lineNumber</code> field set to |lineNumber| and the
+       <code>columnNumber</code> field set to |columnNumber|.
+
+ 1. Append |frame info| to |stack trace|.
+
+ 1. Return |stack trace|
+
+### Events ### {#module-log-events}
+
+#### entryAdded #### {#event-log-entryAdded}
+
+<dl>
+   <dt>Event Type</dt>
+   <dd>
+      <pre class="cddl local-cddl">
+        LogEntryAddedEvent = {
+         method: "log.entryAdded",
+         params: (
+             GenericLogEntry //
```

Comment:
Nit: Could we define a LogEntry production for this grouping up above and reference LogEntry here? That way the list of supported LogEntry types appears closer to their actual definitions.